### ISSUE TITLE
Views hijack focus from all seats on map

### ISSUE THREAD

**Issue by YaLTeR** [swaywm/sway#2918]
`sway version 1.0-beta.1-2-ga4d68358 (Oct 21 2018, branch 'master')`
[sway.log](https://gist.github.com/YaLTeR/f2187b8ca50c48ac7de7e38c940403b7#file-sway-log)
[config](https://gist.github.com/YaLTeR/f2187b8ca50c48ac7de7e38c940403b7#file-config)

I was testing multiple seats and found that a second seat keyboard behaves very strangely. Inputting mod+enter on it (my sway shortcut for spawning a terminal) spawns two terminals. Opening a wayland client and typing something types every character twice. Opening a second wayland client, switching focus to it and then typing results in the characters being typed into both windows (one character into each).

All of this was happening when typing on the second seat keyboard; the first seat keyboard didn't have any weird behaviors. Also typing into xwayland clients doesn't have any weird behaviors.

In the sway.log session I started sway, pressed mod+enter on the second seat keyboard which spawned two terminals (instead of one), then I exited sway.

**Comment by emersion** [comment]
Maybe the keyboard is added to both seats?

**Comment by YaLTeR** [comment]
Hmm, perhaps. I changed the seat lines to:
```
seat seat0 fallback true
seat seat1 attach "1133:49941:Logitech_Logitech_USB_Keyboard"
```
Then reloaded config, tried opening the terminal and closing it and got a [crash](http://sprunge.us/08zQjM), and then while trying to paste that to a gist got [another crash](http://sprunge.us/wjPiC6).

After reloading the session, I tried opening two wayland clients and while switching focus from seat0 resulted in me being able to correctly type in different clients with different keyboards, trying to bring focus back and then switch it from seat1 resulted in the same behavior (so seat0 typing in the newly focused client and seat1 in the old client, while it should've been the other way around).

While typing the rest of this issue from my primary seat0 keyboard with seat1 keyboard just attached and not doing anything, I got a [third crash](http://sprunge.us/ZwZULL). I later got the same crash with seat1 keyboard not even attached.

**Comment by emersion** [comment]
Does https://github.com/swaywm/sway/pull/3390 fix this issue?

**Comment by YaLTeR** [comment]
Yeah, these issues seem to have been fixed on the latest master. There's one more thing I'd classify as unintuitive behavior: opening a client redirects all seats' inputs into it rather than just the seat that opened the client.

**Comment by emersion** [comment]
Yeah, I'm not sure how to fix this. We can't really guess which seat started a client. Maybe we should only focus the default seat?

**Comment by Drakulix** [comment]
> Yeah, I'm not sure how to fix this. We can't really guess which seat started a client. Maybe we should only focus the default seat?

I am not sure, if this a linux specific thing, but since it is possible to figure out the pid of the process, that connects to the wayland socket, would it not be possible to infer, which seat caused a program to start in many cases?

For cases, where sway's exec command is used this should be possible and for other program launches it might be possible to walk up the process hierarchy up to the first process, that was launched by sway exec or that has currently a focused window inside sway.

Maybe this mapping of pids to seats could be cached inside sway to make the lookup easier.

Carrying this information from the execution of a program through to the focus code might be quite difficult (I don't know much about sway's code structure), but technically this should be doable or am I missing something?

**Comment by YaLTeR** [comment]
For sway exec it should be easy to figure out the seat, and for regular programs spawning popups or windows maybe simply take the last seat to interact with the said program?

**Comment by emersion** [comment]
>For sway exec it should be possible to figure out the seat,  and for regular programs spawning popups or windows maybe simply take  the last seat to interact with the said program?

Aye, that should work in most situations.

**Comment by emersion** [comment]
The new xdg-activation protocol can also help with this.

**Comment by optlink** [comment]
> There's one more thing I'd classify as unintuitive behavior: opening a client redirects all seats' inputs into it rather than just the seat that opened the client.

Are there any known workarounds for this? That behavior makes multiseat completely broken in my use case.

**Comment by emersion** [comment]
The workaround is to plumb xdg-activation in Sway and in clients.
